samples: fix GPIO API usage issues

gmarull opened the pull request:
- Add missing CONFIG_GPIO=y
- Add/remove header if used/not used

A tiny step towards the removal of `CONFIG_GPIO=y` from all boards.

gmarull committed:
samples: add missing CONFIG_GPIO=y

All samples using the GPIO API must enable the CONFIG_GPIO Kconfig
option.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: fix zephyr/drivers/gpio.h usage

Only include the header if using any of the GPIO API calls/
struct/defines.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

kartben commented on samples/subsys/tracing/prj.conf:
```diff
@@ -5,3 +5,4 @@ CONFIG_MP_MAX_NUM_CPUS=1
 # Debugging
 CONFIG_DEBUG_OPTIMIZATIONS=n
 CONFIG_DEBUG_THREAD_INFO=y
+CONFIG_GPIO=y
```
drop -- this is only for GPIO tracing AFAICT and enabled in prj_gpio.conf (https://docs.zephyrproject.org/latest/samples/subsys/tracing/README.html#usage-for-gpio-tracing-backend).

Can you maybe double check whatever grep logic you used to ensure it's only for sample actually using gpio API, and in their main.c, that you enable the Kconfig? I suspect this particular sample might not be the only one where this happened.

Thalley commented on samples/bluetooth/iso_receive/prj.conf:
```diff
@@ -5,3 +5,4 @@ CONFIG_LOG=y
 CONFIG_BT_DEVICE_NAME="Test ISO Receive"
 
 CONFIG_BT_ISO_MAX_CHAN=2
+CONFIG_GPIO=y
```
This PR is doing something similar to what https://github.com/zephyrproject-rtos/zephyr/pull/81435 attempted, and my review comment is the same:
The GPIO usage in many/most of these samples are optional, i.e. it most cases I am aware of, it checks if there is an LED and if there is, it toggles that. 

Enabling GPIO in the prj.conf file will provide build warnings/errors for boards that does not have that, even though the boards can run the sample without the LED toggle just fine. 

In other words, GPIO is optional in many of these samples, and this is effectively making it mandatory. 

I'm not familiar with all samples, and some may actually require it, but the `iso_receive` and `periodic_sync` samples only use LED if it is there, and if there's no LED there's no GPIO usage. 

@cvinayak I think you've added this for some of these samples. What are your thoughts?

gmarull replied:
For such cases, the sample should:

- Have a Kconfig option that selects GPIO if LED exists and enabled in DT
- Or, have an alternate config (e.g. in prj_gpio.conf) that can be used to test with GPIO enabled

Which one is preferred?

gmarull replied:
> > if a board has a LED defined, build will fail if it doesn't have CONFIG_GPIO=y, so this is wrong.
> 
> Is it possible that `DT_NODE_HAS_STATUS_OKAY(LED0_NODE)` returns `1` without GPIO? :o
> 

Yes, this is just checking DT. A LED node could implement other APIs, in fact, e.g. LED API.

> In which case, the check should be `DT_NODE_HAS_STATUS_OKAY(LED0_NODE) && IS_ENABLED(CONFIG_GPIO)`.

That works for me. I'll update the PR.

Thalley replied:
> That works for me. I'll update the PR.

Great - But we will need to determine where/how to enable GPIO :)

gmarull replied:
please check latest patch set

Thalley replied:
Now it's back to only being enable by the board itself, isn't it? So once GPIO is removed from boards, it's not longer automatically enabled for board that support the feature? This seems contrary to what the purpose of this PR is supposed to do

gmarull replied:
I've added a sample variant that enables GPIO, isn't that sufficient? If not you need to create an auxiliary Kconfig option that defaults to y if DT has a LED node enabled, selects GPIO and is used as a condition to compile that GPIO branches.

gmarull replied:
@Thalley check latest patch set, I think it is better.

Thalley replied:
> @Thalley check latest patch set, I think it is better.

Yeah, that looks good to me

Should that be applied to all samples where you've set `CONFIG_GPIO=y`?

gmarull replied:
Some samples use GPIO unconditionally, so not always. It's true that we could probably find some common patterns and simplify some samples.

gmarull committed:
samples: bluetooth: iso_receive: add option to control led0

Add a new Kconfig option to control the behavior of led0. This option
also allows to select GPIO only when needed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: bluetooth: mesh_provisioner: add option to control sw0

Add a new Kconfig option to control the behavior of sw0. If enabled, sw0
button will be used for provisioning. This option also allows to select
GPIO.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: bluetooth: periodic_sync: add option to control led0

Add a new option to control led0 behavior. If enabled, led0 will be used
for blinking. This option also allows to select GPIO when needed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

Thalley commented on samples/boards/espressif/deep_sleep/prj.conf:
```diff
@@ -1 +1,2 @@
 CONFIG_POWEROFF=y
+CONFIG_GPIO=y
```
Looks like this seem also uses GPIO optionally. The sample already has 2 Kconfig options, and only if they are enabled is GPIO used as far as I can tell. Should be trivial to add `select GPIO` to those

gmarull replied:
fixed

Thalley commented on samples/drivers/espi/prj.conf:
```diff
@@ -1,3 +1,4 @@
 CONFIG_ESPI=y
 CONFIG_STDOUT_CONSOLE=y
 CONFIG_PRINTK=y
+CONFIG_GPIO=y
```
Looks like this sample also optionally uses GPIO

gmarull replied:
addressed

Thalley commented on samples/drivers/i2s/echo/prj.conf:
```diff
@@ -1 +1,2 @@
 CONFIG_I2S=y
+CONFIG_GPIO=y
```
Optional GPIO

gmarull replied:
addressed

Thalley commented on samples/subsys/display/lvgl/prj.conf:
```diff
@@ -17,3 +17,4 @@ CONFIG_LV_USE_ARC=y
 CONFIG_LV_USE_IMG=y
 CONFIG_LV_USE_MONKEY=y
 CONFIG_LV_FONT_MONTSERRAT_14=y
+CONFIG_GPIO=y
```
Optional GPIO

gmarull replied:
addressed

Thalley reviewed (commented):
Changes LGTM

Some samples seem to optionally support GPIO but still has `GPIO=y`, but those are outside my area, so not sure whether these should be updated. 

Will await with approval a bit

gmarull committed:
samples: subsys: display: lvgl: add option to control sw0

Add a new option to control the behavior of sw0. This option also allows
to select GPIO when needed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: i2s: echo: add options to control sw0/sw1

Add options to control the behavior of sw0/sw1. These options also allow
to select GPIO when needed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: espi: add option to control board_power

Add a new option to control usage of DT nodelabel board_power. This
option also allows to select GPIO when needed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: boards: espressif: deep_sleep: add missing select GPIO

Kconfigs enabled functionality that requires GPIO, but they did not
select GPIO.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>
